presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java

```diff
@@ -0,0 +1,147 @@
+import com.facebook.presto.testing.TestingConnectorSession;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public interface BaseMetadataTest
+        extends SPITest
+{
+    default List<String> systemSchemas()
+    {
+        return ImmutableList.of();
+    }
+
+    default SchemaTableName schemaTableName(String tableName)
+    {
+        return new SchemaTableName("default_schema", tableName);
+    }
+
+    default SchemaTableName schemaTableName(String schemaName, String tableName)
+    {
+        return new SchemaTableName(schemaName, tableName);
+    }
+
+    @Test
+    default void testEmptyMetadata()
+    {
+        ConnectorSession session = new TestingConnectorSession(ImmutableList.of());
+
+        run(this,
+                ImmutableList.of(
+                        metadata -> assertEquals(metadata.listSchemaNames(session), systemSchemas()),
+                        metadata -> assertEquals(metadata.listTables(session, null), ImmutableList.of())));
+    }
+
+    default List<String> schemaNamesOf(SchemaTableName... schemaTableNames)
+    {
+        return Arrays.stream(schemaTableNames)
+                .map(SchemaTableName::getSchemaName)
+                .collect(toImmutableList());
+    }
+
+    default List<String> distinctSchemas(SchemaTableName... schemaTableNames)
+    {
+        return schemaNamesOf(schemaTableNames).stream()
+                .distinct()
+                .collect(toImmutableList());
+    }
+
+    default SchemaTablePrefix prefixOfSchemaName(SchemaTableName schemaTableName)
+    {
+        return new SchemaTablePrefix(schemaTableName.getSchemaName());
+    }
+
+    default SchemaTablePrefix prefixOf(SchemaTableName schemaTableName)
+    {
+        return new SchemaTablePrefix(schemaTableName.getSchemaName(), schemaTableName.getTableName());
+    }
+
+    default void run(SPITest test, List<Consumer<ConnectorMetadata>> consumers)
+    {
+        consumers.forEach(consumer -> withMetadata(test, ImmutableList.of(consumer)));
+    }
+
+    default void withMetadata(SPITest test, List<Consumer<ConnectorMetadata>> consumers)
+    {
+        Connector connector = test.getConnector();
+        ConnectorTransactionHandle transaction = connector.beginTransaction(IsolationLevel.READ_UNCOMMITTED, true);
+        ConnectorMetadata metadata = connector.getMetadata(transaction);
+        consumers.forEach(consumer -> consumer.accept(metadata));
+        connector.commit(transaction);
+    }
+
+    default List<Consumer<ConnectorMetadata>> withSchema(ConnectorSession session, List<String> schemaNames, List<Consumer<ConnectorMetadata>> consumers)
```

Comment:
Unfortunate side-effect of having this in an interface: Can't have default final methods, meaning we can't annotate these @SafeVarargs to suppress the warning that otherwise results.